docs/nima/testing.adoc

```diff
@@ -0,0 +1,397 @@
+
+Routing is configured the same way as in full server testing using the `@SetUpRoute` annotation.
+
+== Additional Information
+
+=== WebSocket Testing
+
+In case WebSocket testing is required, there is an additional module for this. It is necessary to include the following Maven dependency to Project's pom file:
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.helidon.nima.testing.junit5</groupId>
+    <artifactId>helidon-nima-testing-junit5-websocket</artifactId>
+    <scope>test</scope>
+</dependency>
+----
+
+
+=== WebSocket Testing Example
+
+WebSocket Testing extension adds support for routing configuration and injection of WebSocket related artifacts, such as `WebSockets` as well `DirectWsClient` in Helidon Níma unit tests.
+
+.WebSocket sample test.
+[source,java]
+----
+@ServerTest
+class WsSocketTest {
+
+    private static final ServerSideListener WS_LISTENER = new ServerSideListener();
+
+    private final WsClient wsClient;    <1>
+
+    protected WsSocketTest(WsClient wsClient) {
+        this.wsClient = wsClient;
+    }
+
+    @SetUpRoute
+    static void routing(WsRouting.Builder ws) { <2>
+        ws.endpoint("/testWs", WS_LISTENER);
+    }
+
+    @Test
+    void testWsEndpoint() { <3>
+        WS_LISTENER.reset();
+
+        ClientSideListener clientListener = new ClientSideListener();
+        wsClient.connect("/testWs", clientListener);    <4>
+
+        assertThat(clientListener.message, is("ws"));   <5>
+    }
+}
+----
+<1> Declare `WsClient` and later inject it in the constructor;
+<2> Using `@SetUpRoute` create WebSocket routing and assign a serverside listener;
+<3> Test the WebSocket endpoint using regular `@Test` annotation;
+<4> Create and assign clientside listener;
+<5> Check if the received message is correct;
+
+
+FIXME: not sure we need this, maybe as popup?
+
+WebSocket `ClientSideListener` is a helper class implementing `WsListener`, is very simple and looks this way:
+
+.`ClientSideListener` helper class.
+[source,java]
+----
+private static class ClientSideListener implements WsListener {
+    private final CountDownLatch cdl = new CountDownLatch(1);
+    private String message;
+    private volatile Throwable throwable;
+
+    @Override
+    public void onOpen(WsSession session) { <1>
+        session.send("hello", true);
+    }
+
+    @Override
+    public void onMessage(WsSession session, String text, boolean last) {   <2>
+        this.message = text;
+        session.close(WsCloseCodes.NORMAL_CLOSE, "End");
+    }
+
+    @Override
+    public void onClose(WsSession session, int status, String reason) { <3>
+        cdl.countDown();
+    }
+
+    @Override
+    public void onError(WsSession session, Throwable t) {   <4>
+        this.throwable = t;
+        cdl.countDown();
+    }
+}
+----
+<1> Send "Hello", when a connection is opened;
```

Comment:
Better to end sentences with a period.
<1> Send "Hello" when a connection is opened. (removed the comma)

Reply:
ok)